src/mjolnir.py

```diff
@@ -0,0 +1,220 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the Mjolnir class to represent the Mjolnir plugin for Synapse."""
+
+import logging
+import typing
+from secrets import token_hex
+
+import ops
+from ops.jujuversion import JujuVersion
+
+import actions
+import synapse
+from charm_state import CharmState
+from constants import (
+    MJOLNIR_MANAGEMENT_ROOM,
+    MJOLNIR_MEMBERSHIP_ROOM,
+    MJOLNIR_SERVICE_NAME,
+    PEER_RELATION_NAME,
+    SECRET_ID,
+    SECRET_KEY,
+    SYNAPSE_CONTAINER_NAME,
+)
+from user import User
+
+logger = logging.getLogger(__name__)
+
+MJOLNIR_USER = "mjolnir"
+
+
+class Mjolnir(ops.Object):  # pylint: disable=too-few-public-methods
+    """A class representing the Mjolnir plugin for Synapse application."""
+
+    def __init__(self, charm: ops.CharmBase, charm_state: CharmState):
```

Comment:
The charm because Mjolnir uses the peer data/secrets to store an admin access token that is used to further actions.
The charm_state because it checks if enable_mjolnir configuration is set in order to configure the plugin or not.

Reply:
@gregory-schiano this is an interesting case, I think the peer data/ secret should come from the charm state because it is another source of data that the charm needs to function, much like relation data

Reply:
Alright, just laying out my thoughts here so I can see whether I understand:
The mjolnir class now has both observer and business logic and we want to separate these. The class is rewritten to be an observer, the business logic lives separately in a service module (maybe for all future plugins as well?), the state is populated with the necessary data, so it lives in the state and the observer reconciles on changes. Is this correct?

Reply:
correct

Reply:
In order to provide this feature to the Community team, I suggest that we work on this change in a follow-up PR. WDYT?

IMHO since it's more related to an architecture definition, in a follow-up PR we could have more discussion and it would be easier for other team members to see the changes and discuss them as well.

Reply:
ok sounds good